File: java/RekeningBank.java
/**
 * Sesi 3 — constructor berdelegasi, anggota statis, dan konstanta.
 *
 * Invariant:
 *   1. saldo tidak pernah negatif
 *   2. nomor rekening tidak berubah setelah objek dibuat
 *   3. setoran dan penarikan selalu bernilai positif
 */
public class RekeningBank {

    // TODO 1 -> konstanta bernama (selesai)
    public static final double BUNGA_TAHUNAN = 0.025;
    public static final double BIAYA_ADMIN = 5000;
    public static final double BATAS_PENARIKAN = 5_000_000;

    // TODO 2 -> penghitung jumlah rekening (selesai)
    private static int jumlahRekening = 0;

    private final String nomor;
    private final String pemilik;
    private double saldo;

    /**
     * Constructor ringkas.
     * TODO 3 -> delegasi ke constructor lengkap (selesai)
     */
    public RekeningBank(String nomor, String pemilik) {
        this(nomor, pemilik, 0);   // delegasi: validasi & increment counter cukup di constructor lengkap
    }

    /** Constructor lengkap — SATU-SATUNYA tempat validasi berada. */
    public RekeningBank(String nomor, String pemilik, double saldoAwal) {
        // TODO 4 -> validasi (selesai)
        if (nomor == null || nomor.isBlank()) {
            throw new IllegalArgumentException("Nomor rekening tidak boleh kosong");
        }
        if (saldoAwal < 0) {
            throw new IllegalArgumentException("Saldo awal tidak boleh negatif");
        }

        this.nomor = nomor;
        this.pemilik = pemilik;
        this.saldo = saldoAwal;

        // TODO 5 -> naikkan counter DI SINI SAJA (selesai)
        jumlahRekening++;
    }

    public void setor(double jumlah) {
        // TODO 6 -> selesai
        if (jumlah <= 0) {
            throw new IllegalArgumentException("Jumlah setoran harus lebih dari 0");
        }
        this.saldo += jumlah;
    }

    public void tarik(double jumlah) {
        // TODO 7 -> selesai
        if (jumlah <= 0) {
            throw new IllegalArgumentException("Jumlah penarikan harus lebih dari 0");
        }
        if (jumlah > BATAS_PENARIKAN) {
            throw new IllegalArgumentException("Melebihi batas penarikan sekali transaksi");
        }
        if (jumlah > saldo) {
            throw new IllegalArgumentException("Saldo tidak mencukupi");
        }
        this.saldo -= jumlah;
    }

    /** TODO 8 -> selesai */
    public void potongBiayaAdmin() {
        this.saldo = Math.max(0, this.saldo - BIAYA_ADMIN);
    }

    /** TODO 9 -> selesai */
    public static int getJumlahRekening() {
        return jumlahRekening;
    }

    /** TODO 10 -> selesai (murni pakai parameter, tidak baca state objek -> pantas static) */
    public static double bungaSetahun(double pokok) {
        return pokok * BUNGA_TAHUNAN;
    }

    public double getSaldo()  { return saldo; }
    public String getNomor()  { return nomor; }

    @Override
    public String toString() {
        return String.format(java.util.Locale.US, "Rekening[%s] %-14s Rp%,.2f", nomor, pemilik, saldo);
    }
}

File: php/RekeningBank.php
<?php
declare(strict_types=1);

/**
 * Sesi 3 — RekeningBank (PHP)
 */
class RekeningBank {

    // Konstanta
    public const BUNGA_TAHUNAN = 0.025;
    public const BIAYA_ADMIN = 5000;
    public const BATAS_PENARIKAN = 5_000_000;

    // Property statis
    private static int $jumlahRekening = 0;

    // Property instance
    private string $nomor;
    private string $pemilik;
    private float $saldo;

    /**
     * Constructor (menggunakan default parameter untuk saldo awal = 0)
     */
    public function __construct(string $nomor, string $pemilik, float $saldoAwal = 0.0) {
        if (trim($nomor) === '') {
            throw new InvalidArgumentException("Nomor rekening tidak boleh kosong");
        }
        if ($saldoAwal < 0) {
            throw new InvalidArgumentException("Saldo awal tidak boleh negatif");
        }

        $this->nomor = $nomor;
        $this->pemilik = $pemilik;
        $this->saldo = $saldoAwal;

        self::$jumlahRekening++;
    }

    public function setor(float $jumlah): void {
        if ($jumlah <= 0) {
            throw new InvalidArgumentException("Jumlah setoran harus lebih dari 0");
        }
        $this->saldo += $jumlah;
    }

    public function tarik(float $jumlah): void {
        if ($jumlah <= 0) {
            throw new InvalidArgumentException("Jumlah penarikan harus lebih dari 0");
        }
        if ($jumlah > self::BATAS_PENARIKAN) {
            throw new InvalidArgumentException("Melebihi batas penarikan sekali transaksi");
        }
        if ($jumlah > $this->saldo) {
            throw new InvalidArgumentException("Saldo tidak mencukupi");
        }
        $this->saldo -= $jumlah;
    }

    public function potongBiayaAdmin(): void {
        // Menggunakan 0.0 agar nilai kembalian tetap float
        $this->saldo = max(0.0, $this->saldo - self::BIAYA_ADMIN);
    }

    public static function getJumlahRekening(): int {
        return self::$jumlahRekening;
    }

    public static function bungaSetahun(float $pokok): float {
        return $pokok * self::BUNGA_TAHUNAN;
    }

    public function getSaldo(): float {
        return $this->saldo;
    }

    public function getNomor(): string {
        return $this->nomor;
    }

    public function __toString(): string {
        $formattedSaldo = "Rp" . number_format($this->saldo, 2, '.', ',');
        return sprintf("Rekening[%s] %-14s %s", $this->nomor, $this->pemilik, $formattedSaldo);
    }
}

// KODE EKSEKUSI PROGRAM
echo "Jumlah rekening di awal: " . RekeningBank::getJumlahRekening() . "\n";

$rek1 = new RekeningBank("111", "Ani", 1_000_000);
$rek2 = new RekeningBank("222", "Budi");
$rek3 = new RekeningBank("333", "Citra", 250_000);

echo $rek1 . "\n";
echo $rek2 . "\n";
echo $rek3 . "\n";
echo "Jumlah rekening sekarang: " . RekeningBank::getJumlahRekening() . "   (seharusnya 3, bukan 4)\n\n";

echo "=== Operasi ===\n";
$rek1->setor(500_000);
echo "Setelah setor 500.000 -> " . $rek1 . "\n";

try {
    $rek1->tarik(6_000_000);
} catch (InvalidArgumentException $e) {
    echo "Ditolak: " . $e->getMessage() . "\n";
}

$rek2->potongBiayaAdmin();
echo "Budi setelah potong admin: " . $rek2 . "   (saldo tidak boleh negatif)\n";

$bungaAni = "Rp" . number_format(RekeningBank::bungaSetahun($rek1->getSaldo()), 2, '.', ',');
echo "Bunga setahun dari saldo Ani: " . $bungaAni . "\n";